File: app/src/main/java/net/xzos/UpgradeAll/database/HubDatabase.java
package net.xzos.UpgradeAll.database;

import androidx.annotation.NonNull;

import com.google.gson.Gson;
import com.google.gson.JsonSyntaxException;

import net.xzos.UpgradeAll.gson.HubConfig;
import net.xzos.UpgradeAll.gson.HubItemExtraData;

import org.litepal.annotation.Column;
import org.litepal.crud.LitePalSupport;

public class HubDatabase extends LitePalSupport {
    private int id;
    private String name;
    @Column(unique = true)
    private String uuid;
    private String hub_config;
    private String extra_data;  // JSON形式 存储JavaScript 和 额外数据

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public HubConfig getHubConfig() {
        Gson gson = new Gson();
        try {
            return gson.fromJson(hub_config, HubConfig.class);
        } catch (JsonSyntaxException e) {
            return null;
        }
    }

    public void setHubConfig(HubConfig repo_config) {
        Gson gson = new Gson();
        this.hub_config = gson.toJson(repo_config);
    }

    public String getUuid() {
        return uuid;
    }

    public void setUuid(@NonNull String uuid) {
        this.uuid = uuid.toLowerCase();
    }

    public HubItemExtraData getExtraData() {
        Gson gson = new Gson();
        try {
            return gson.fromJson(extra_data, HubItemExtraData.class);
        } catch (JsonSyntaxException e) {
            return null;
        }
    }

    public void setExtraData(HubItemExtraData extra_data) {
        Gson gson = new Gson();
        this.extra_data = gson.toJson(extra_data);
    }
}

File: app/src/main/kotlin/net/xzos/upgradeAll/database/HubDatabase.kt
package net.xzos.upgradeAll.database

import com.google.gson.Gson
import net.xzos.upgradeAll.json.gson.HubConfig
import net.xzos.upgradeAll.json.gson.HubDatabaseExtraData
import org.litepal.crud.LitePalSupport

open class HubDatabase(
        var name: String,
        var uuid: String,
        private var hub_config: String?,
        private var extra_data: String?
) : LitePalSupport() {
    val id: Long = 0

    var hubConfig: Any?
        set(value) {
            if (value != null && value is HubConfig)
                hub_config = Gson().toJson(value)
        }
        get() {
            return if (hub_config != null) Gson().fromJson(hub_config, HubConfig::class.java)
            else null
        }
    var extraData: Any?
        set(value) {
            if (value != null && value is HubDatabaseExtraData)
                extra_data = Gson().toJson(value)
        }
        get() {
            return if (extra_data != null) Gson().fromJson(extra_data, HubDatabaseExtraData::class.java)
            else null
        }
}